syntax = "proto3";

package orderbook;

service OrderBookService {
  rpc SubscribeOrderBook (stream SubscriptionRequest) returns (stream OrderBookUpdate) {}
  rpc PlaceOrder (Order) returns (OrderResponse) {}
}

message SubscriptionRequest {
  string symbol = 1;
  bool subscribe = 2;
}

message OrderBookUpdate {
  string symbol = 1;
  repeated PriceLevel bids = 2;
  repeated PriceLevel asks = 3;
  bool is_snapshot = 4;
  repeated PriceLevelUpdate changes = 5;
  int64 version = 6;
}

message PriceLevel {
  string price = 1;
  int32 quantity = 2;
}

message PriceLevelUpdate {
  string price = 1;
  int32 quantity = 2;
  Side side = 3;
  Action action = 4;
}

enum Side {
  BID = 0;
  ASK = 1;
}

enum Action {
  ADD = 0;
  UPDATE = 1;
  DELETE = 2;
}

message Order {
  string symbol = 1;
  string order_id = 2;
  string side = 3;
  string type = 4;
  string price = 5;
  int32 quantity = 6;
}

message OrderResponse {
  string order_id = 1;
  string status = 2;
}